### DISCUSSION TITLE
COS-Tools and Andras Tantos’s Cray Simulator Question

### DISCUSSION THREAD

**Discussion by nobunagashino** [kej715/COS-Tools#2] [Q&A]
Hi,
I have been experimenting (1) with COS-Tools and Andras Tantos’s Cray Simulator and think I have
Found a few problems with the CAL assembler and maybe the link loader. 
(1)	The hello.sh causes the cal program to hang immediately taking almost 100% of the cpu and producing no output.
(2)	I have a small test program,test.cal which assembles and links without any errors. Once I try to run it in batch or interactive mode (2) an “invalid program load table” error is reported and the  program aborts.

So to be honest I don’t know if I am doing something stupid or there really is a problem with the cal/ldr programs. Any Advice/hints/tips welcome.
To assist I will try to upload the following files.
Test.txt  --the assembler source of my tiny test program. 
Cray-assembly.jpg shows the assembler command line and the ldr output.
Cray-run-result.jpg shows the error output from an interactive job. A Batch job produces the same error.
(1)	Feels a bit like blundering around in the dark-I have never used a mainframe style computer so quite a learning curve. Fun though..
(2)	I use the build_exp_disk script to load the executable on the virtual disk which can then be fetched into the CPU.
(3)	You might ask why I am bothering with this? As a child of the late 1960’s CRAY computers have held something of a special spot(4). So now that I can finally get my paws on one I really want to write a program to run 2+2 and print the result.. I would also suspect ,on a more serious note, that there are computer science students out there who find this of interest. Might be an interesting way to learn assembly code. 
(4)	Personally, I have only seen CRAY computers in the science museum-my late father once worked on one which was in Wythenshawe Manchester owned by SHELL.

![cray-assembly](https://user-images.githubusercontent.com/46041147/230626547-79a4c5db-a214-49bf-9ea2-bee1e5d28269.jpg)
![cray-assembly-file-size](https://user-images.githubusercontent.com/46041147/230626554-b9daf20c-e1d5-4c49-afcc-84e1957734fc.jpg)
![cray-run-result](https://user-images.githubusercontent.com/46041147/230626557-d513edd5-449f-461d-b5d7-ad7eee9c8801.jpg)
[test.txt](https://github.com/kej715/COS-Tools/files/11179549/test.txt)

**Comment by kej715** [comment]
Thanks for reporting the problem with hello.sh. Code has been pushed to resolve the issue.

Please provide a copy of your test program. Code produced by cal and ldr has been tested successfully in an environment where DtCyber serves as a frontend to COS 1.17. Binaries produced by ldr can be uploaded to DtCyber running NOS 2.8.7, then transmitted to COS 1.17 using the Cray Station interface. This preserves the Cray file structure created by ldr, and the binaries load and run successfully. I suspect that build_exp_disk does not preserve the file structure properly. For example, perhaps it wraps it unnecessarily in a second layer of file structure.

If you provide a copy of your test program, I can try running it using the Cray Station interface.

**Comment by kej715** [comment]
I had overlooked the attachment of your test program earlier. After seeing it, I assembled and loaded it and then uploaded the executable to host NCCMAX, a CDC Cyber 865 at the Nostalgic Computing Center (http://www.nostalgiccomputing.org). NCCMAX is connected via a Cray Station interface to a Cray X-MP running COS 1.17. The test program loaded successfully and ran, but it aborted because it does not include an operating system call to terminate gracefully. Consequently, an illegal instruction is encountered after the "S2 S2+S3" instruction at the end of the program. When I modified the program to add a call to the "ADV" macro after "S2 S2+S3", the program ran to completion and exited normally. The "ADV" macro executes an OS call to advance to the next job step, i.e., the usual way that a COS program signals normal completion.

Here is the job output after adding the ADV macro:

```
  06:59:37.5975       0.0007    CSP             CRAY XMP-14 SN 302      LEADING EDGE TECHNOLOGIES       04/08/83
  06:59:37.6004       0.0008    CSP
  06:59:37.6039       0.0009    CSP             CRAY OPERATING SYSTEM           COS 1.17  ASSEMBLY DATE 02/28/89
  06:59:37.6081       0.0010    CSP
  06:59:37.6107       0.0011    CSP
  06:59:37.6371       0.0013    CSP     JOB,JN=CRAYRUN.
  06:59:37.7423       0.0105    CSP     ACCOUNT,AC=,APW=,UPW=.
  06:59:37.7465       0.0110    CSP
  06:59:37.7493       0.0111    CSP     .......................................................................
  06:59:37.7522       0.0112    CSP
  06:59:37.7923       0.0123    CSP     ECHO,ON=ALL.
  06:59:37.8026       0.0130    CSP     OPTION,STAT=ON.
  06:59:37.8171       0.0160    CSP     FETCH,DN=BIN,MF=FE,DF=TR,^
  06:59:37.8239       0.0190    CSP     TEXT=.
  06:59:42.2974       0.0193    SCP     AAHR 08.36.26.USER,Z160000,.
  06:59:42.2994       0.0193    SCP     AAHR 08.36.26.CHARGE,   *                  ,
  06:59:42.3013       0.0193    SCP     AAHR 08.36.29. TRO - PF AND TAPE TRANSPARENT OUTPUT.
  06:59:42.3033       0.0193    SCP     AAHR 08.36.30.TRO: DATASET BIN    ,         63B WORDS,     0.277
  06:59:42.6143       0.0193    SCP     SS004 - DATASET RECEIVED FROM FRONT END
  06:59:43.9493       0.0210    CSP     BIN.
  06:59:43.9828       0.0216    CSP     END OF JOB
  06:59:43.9877       0.0219    EXP     SY005 - BIN           512 WRDS,    1 IOS,      1 REQ,       1 STRS,     .019 SEC
  06:59:43.9901       0.0219    EXP           - 29-1-32A       18 STRS   READ:         1 REQ,       1 STRS,     .021 SEC
  06:59:43.9997       0.0220    CSP
  06:59:44.0022       0.0221    CSP
  06:59:44.0068       0.0224    CSP     CHARGES
  06:59:44.0145       0.0225    CSP     CS032 - DATASET NOT LOCAL:
  06:59:44.0182       0.0226    ABORT   AB025 - USER PROGRAM REQUESTED ABORT
  06:59:44.0204       0.0226    ABORT         - P=00001417B   TASK-ID=0001
  06:59:44.0228       0.0226    ABORT         - BASE=00775000 LIMIT=01153000 CPU-NUMBER=00
  06:59:44.0250       0.0226    ABORT         - JOB STEP ABORTED.
```

Here is the NOS procedure file used for executing the COS executable. The executable is uploaded to the NOS host in binary mode using Kermit:

```
.PROC,CRAYRUN,F=CRAYBIN.
REWIND,F.
COPYBF,F,ZZZCBIN.
REPLACE,ZZZCBIN=CSTAGE.
CSUBMIT,CRAYJOB,TO.
REVERT,NOLIST.
.DATA,CRAYJOB.
JOB,JN=CRAYRUN.
ACCOUNT,AC=CRAY,APW=XYZZY,UPW=QUASAR.
ECHO,ON=ALL.
OPTION,STAT=ON.
FETCH,DN=BIN,MF=FE,DF=TR,^
TEXT='USER,nosuser,nospass,BIN=CSTAGE.CTASK.'.
BIN.
```

Here is the source code of the version of test.txt that ran to completion successfully:

```
         TITLE     'TEST'
         SUBTITLE  'TEST'
         IDENT     TEST
         COMMENT   'TEST'
         ENTRY     TEST
         EXT       ADV
         START     TEST
TEST     S1        =*
         S2        S2+S3           ; CLASS=1, OR=1, FC=3
         R         ADV             ; advance to next job step

         END
```

**Comment by nobunagashino** [comment]
Hi Kevin,

Thanks once again for looking at this-delighted it worked. On my end I 
have DtCyber compiled,NOS2.8.7 up and

working and a test site.cfg to connect to Tanto's simulator.

Going to take a day or two looking at the manual and  docs on 
bitsavers-it's quite a learning curve!

Best wishes,

Ian

**Reply by kej715** [reply]
Hi Ian,

It's great to read that you've installed DtCyber with NOS 2.8.7 and are planning to use it as a frontend to your Cray X-MP running COS 1.17.  FYI ... the simplest site.cfg for enabling the Cray Station interface would look like:

```
[NETWORK]
crayStation=MyXMP,XMP,127.0.0.1:9000
```

This assumes that your X-MP is running on the same host machine as DtCyber. Port 9000 is the default port used by the Cray Station FEI on the X-MP side, so there should be no need to modify the config file for Tantos' emulator. Similarly, "FE" is the default frontend identifier expected by COS 1.17, and "C1" is the default Cray mainframe identifier.  So, if you've run the "reconfigure" tool to enable and activate the Cray Station interface on the DtCyber side, the two machines should connect automatically to each other after deadstarting both. Nothing special needs to be done on the Cray side other than executing the STATION command to activate the station interface, which is a normal part of starting COS 1.17 according to Tantos' wiki.

On the NOS 2.8.7 side, `CSUBMIT` is the command for submitting jobs to the Cray X-MP. The CCL procedure `CRAYRUN` given above, and repeated here, can be saved in your GUEST and/or INSTALL account. It uses CSUBMIT to submit a job to COS that fetches an executable saved on the NOS side and then runs it. The executable can be produced by the `cal` and `ldr` tools and uploaded to NOS using Kermit or FTP. Make sure to use binary mode when using these tools to upload an executable produced by `ldr`.

`CRAYRUN` specifies the `TO` parameter when it calls `CSUBMIT`. This causes the output of the COS job to be returned to the wait queue on NOS. You can use the `STATUS,JSN` command on NOS to wait for the output to arrive, and you can use the `QGET` command to retrieve it from the queue when it does arrive.

```
.PROC,CRAYRUN,F=CRAYBIN.
REWIND,F.
COPYBF,F,ZZZCBIN.
REPLACE,ZZZCBIN=CSTAGE.
CSUBMIT,CRAYJOB,TO.
REVERT,NOLIST.
.DATA,CRAYJOB.
JOB,JN=CRAYRUN.
ACCOUNT,AC=CRAY,APW=XYZZY,UPW=QUASAR.
ECHO,ON=ALL.
OPTION,STAT=ON.
FETCH,DN=BIN,MF=FE,DF=TR,^
TEXT='USER,GUEST,GUEST,BIN=CSTAGE.CTASK.'.
BIN.
```

Replace username and password `GUEST`, above, with `INSTALL` if you upload executables to the INSTALL account instead of the GUEST one.

Assuming that you've SAVEd CRAYRUN on GUEST or INSTALL, and that you've used FTP to upload a COS executable named HELLO, you would run the executable using the following commands:

```
GET,CRAYRUN  (only needs to be executed once after login)
GET,HELLO
CRAYRUN,F=HELLO
```

then use `STATUS,JSN` to monitor the queues, and when output from the COS job arrives in the WAIT queue, use `QGET,jsn` to retrieve it. You can then use `LIST,F=jsn` to view it, or `ROUTE,jsn,DC=PR` to send it to the line printer.

cheers,
Kevin

**Reply by kej715** [reply]
P.S.:  The Cray Station subsystem on NOS 2.8.7 includes the `ICF` network application (Interactive Cray Facility). This allows you to log into COS interactively from NOS. To use it, log into NOS, then execute the command `HELLO,ICF`. This will disconnect your terminal from the NOS interactive facility (IAF) and connect it to the Cray Interactive Facility (ICF). You can then interact with COS in the same way that you would use interactive mode according to Tantos' "Cray Files" discussion.

**Comment by nobunagashino** [comment]
Hi Kevin,

Thanks for the mails-I have the two emulators talking together (on the 
DtCyber console on the B,A side

I see a line for crs and it is reporting cray cos 1.17 02/28/89).On the 
cray side the Link command shows a connection

with the FE id.So all looks happy.

  However.....

when I run /HELLO,ICF I get an application not present as per the 
screenshot. Running /LIMITS shows

that ICF is a valid applicaion for the guest user.Using the INSTALL user 
also produces an "invalid application"

message. When I use help to produce a list of commands ICF is not listed...

Anyway,will  try a batch job as see what happens and have a look in the 
docs to see if there is any clue as

to why ICF appears to be missing.

Once again I really appreciate  your help and time.

Best wishes,

Ian

**Comment by nickglazzard1959** [comment]
Hi folks,
I found I needed a slight variation of the CRAYRUN CCL for it to work. This seems to do things for me:
```
.PROC,CRF,F=CRAYBIN.
REWIND,F.
COPYBF,F,ZZZCBIN.
REPLACE,ZZZCBIN=CSTAGE.
CSUBMIT,CRAYJOB,TO.
REVERT,NOLIST.
.DATA,CRAYJOB.
JOB,JN=CRAYRUN.
ACCOUNT,AC=CRAY,APW=XYZZY,UPW=QUASAR.
ECHO,ON=ALL.
OPTION,STAT=ON.
FETCH,DN=BIN,MF=FE,DF=TR,^
TEXT='USER,nosuser,nospw.GET,BIN=CSTAGE.CTASK.'.
BIN.
```
Note the full stop (period) after the NOS password and the addition of GET to get the file to be local. With this, I get the following output on my 865/XMP system:
```
  18:47:04.6753       0.0002    CSP             CRAY XMP-14 SN 302      LEADING EDGE TECHNOLOGIES       05/02/83
  18:47:04.6763       0.0002    CSP
  18:47:04.6774       0.0003    CSP             CRAY OPERATING SYSTEM           COS 1.17  ASSEMBLY DATE 02/28/89
  18:47:04.6784       0.0004    CSP
  18:47:04.6792       0.0004    CSP
  18:47:04.6885       0.0005    CSP     JOB,JN=CRAYRUN.
  18:47:04.7184       0.0036    CSP     ACCOUNT,AC=,APW=,UPW=.
  18:47:04.7197       0.0038    CSP
  18:47:04.7206       0.0038    CSP     .......................................................................
  18:47:04.7216       0.0039    CSP
  18:47:04.7312       0.0043    CSP     ECHO,ON=ALL.
  18:47:04.7343       0.0046    CSP     OPTION,STAT=ON.
  18:47:04.7388       0.0056    CSP     FETCH,DN=BIN,MF=FE,DF=TR,^
  18:47:04.7408       0.0065    CSP     TEXT=.
  18:47:09.7228       0.0066    SCP     AAHZ 18.48.28.USER,NICK,.
  18:47:09.7238       0.0066    SCP     AAHZ 18.48.28.CHARGE,   *                  ,
  18:47:09.7248       0.0066    SCP     AAHZ 18.48.31. TRO - PF AND TAPE TRANSPARENT OUTPUT.
  18:47:09.7259       0.0066    SCP     AAHZ 18.48.31.TRO: DATASET BIN    ,         71B WORDS,     0.148
  18:47:09.9985       0.0066    SCP     SS004 - DATASET RECEIVED FROM FRONT END
  18:47:10.9748       0.0072    CSP     BIN.
  18:47:10.9793       0.0073    USER    HELLO WORLD!
  18:47:10.9870       0.0075    CSP     END OF JOB
  18:47:10.9886       0.0076    EXP     SY005 - BIN           512 WRDS,    1 IOS,      1 REQ,       1 STRS,     .007 SEC
  18:47:10.9895       0.0076    EXP           - 29-1-22A       18 STRS   READ:         1 REQ,       1 STRS,     .008 SEC
  18:47:10.9923       0.0076    CSP
  18:47:10.9932       0.0076    CSP
  18:47:10.9947       0.0077    CSP     CHARGES
  18:47:10.9961       0.0077    CSP     CS032 - DATASET NOT LOCAL:
  18:47:10.9972       0.0078    ABORT   AB025 - USER PROGRAM REQUESTED ABORT
  18:47:10.9980       0.0078    ABORT         - P=00001417B   TASK-ID=0001
  18:47:10.9988       0.0078    ABORT         - BASE=00775000 LIMIT=01153000 CPU-NUMBER=00
  18:47:10.9996       0.0078    ABORT         - JOB STEP ABORTED.
```
Not sure why there is an abort at the end. Some program COS expects that has sadly been lost?

**Comment by kej715** [comment]
Hi Nick,

Yes, the abort at the end is due to a missing program that would
normally perform accounting at end of job.

cheers,
Kevin

**Comment by kej715** [comment]
... and thanks for pointing out the text (".GET,") that was somehow dropped
from the CRAYRUN procedure example.

**Comment by kej715** [comment]
Thanks to investigation by Nick Glazzard, the reason that ICF was missing
has been determined, and code has been pushed to the DtCyber repository to
resolve the problem. The problem was due to a "leftover" IF statement in
the ICF startup job (leftover from usage in another context). The IF
statement allowed ICF to be started only if it would be running on a host
with machine ID "AX" or "ED".

If you execute "git pull" against your copy of the DtCyber repo and then
reinstall NOS 2.8.7, the problem should be resolved. You can also fix the
problem in your already installed system by editing the NAMSTRT file under
username NETOPS. The very last record in the file is the ICF startup job.
Remove the "IF" statement that tests for HID being either "AX" or "ED", and
then save the file. For example, log into username INSTALL, and then
execute this sequence of commands:

GET,NAMSTRT/UN=NETOPS

SCREEN,WEBTRM

FSE,NAMSTRT

   locate the IF statement near the end of the file and remove it and its
corresponding ENDIF statement

   then exit from FSE

REPLACE,NAMSTRT/UN=NETOPS


Then execute the following command from DSD (i.e., the NOS console) to
start ICF:

X.NAMI(RS=IC)


cheers,
Kevin
